nncf/torch/graph/patterns.py

```diff
@@ -11,28 +11,105 @@
+import networkx as nx
+import networkx.algorithms.isomorphism as ism
 
-RELU = N(VersionAgnosticNames.RELU) | N('hardtanh')
+NODE_COUNTER = 0
 
-BN = N('batch_norm') | N('batch_norm3d')
 
-POOLING = N('adaptive_avg_pool2d') | N('adaptive_avg_pool3d') | N('avg_pool2d') | N('avg_pool3d')
+class GraphPattern:
+    def __init__(self, pattern_name, types):
+        global NODE_COUNTER
+        self.pattern_name = pattern_name
+        self._graph = nx.DiGraph()
+        self._graph.add_node(NODE_COUNTER, type=types)
+        NODE_COUNTER += 1
 
-NON_RELU_ACTIVATIONS = N('elu') | N('elu_') | N('prelu') | N('sigmoid') | N('gelu')
+    def __add__(self, other):
+        def _add_subgraph_to_graph(graph: nx.DiGraph, subgraph: nx.DiGraph) -> None:
+            for node in subgraph.nodes:
+                if node in graph.nodes:
+                    assert False
+                graph.add_node(node, type=subgraph.nodes[node]['type'])
+            for edge in subgraph.edges:
+                graph.add_edge(edge[0], edge[1])
 
-ACTIVATIONS = RELU | NON_RELU_ACTIVATIONS
+        def _merge_one_subgraph_to_other(main: nx.DiGraph, other: nx.DiGraph) -> nx.DiGraph:
+            nodes = list(nx.topological_sort(main))
+            last_node_subgraph = nodes[-1]
 
-ANY_BN_ACT_COMBO = BN + ACTIVATIONS | ACTIVATIONS + BN | BN | ACTIVATIONS
+            for node in other.nodes:
+                main.add_node(node, type=other.nodes[node]['type'])
+            for edge in other.edges:
+                main.add_edge(edge[0], edge[1])
 
-SINGLE_OPS = ACTIVATIONS | POOLING | N('mean') | N('layer_norm')
+            nodes = list(nx.topological_sort(other))
+            first_node_other = nodes[0]
 
-ARITHMETIC = N('__iadd__') | N('__add__') | N('__mul__') | N('__rmul__')
+            main.add_edge(last_node_subgraph, first_node_other)
+            return main
 
-ELTWISE_UNIFORM_OPS = BN | RELU | ACTIVATIONS
+        def _create_copy_of_subgraph(subgraph: nx.DiGraph) -> nx.DiGraph:
+            global NODE_COUNTER
+            mapping = {}
+            for node in subgraph.nodes:
+                new_node = NODE_COUNTER
+                mapping[node] = new_node
+                NODE_COUNTER += 1
+            return nx.relabel_nodes(subgraph, mapping, copy=True)
 
-MATMUL = N('matmul') | N('bmm')
+        weakly_self_subgraphs = self.get_weakly_connected_subgraphs()
+        weakly_other_subgraphs = other.get_weakly_connected_subgraphs()
+        res_graph = nx.DiGraph()
+        for self_subgraph in weakly_self_subgraphs:
+            for other_subgraph in weakly_other_subgraphs:
+                # Create a copy of subgraph and add to a graph
+                subgraph_copy = _create_copy_of_subgraph(self_subgraph)
+                other_subgraph_copy = _create_copy_of_subgraph(other_subgraph)
+                subgraph_copy = _merge_one_subgraph_to_other(subgraph_copy, other_subgraph_copy)
+                _add_subgraph_to_graph(res_graph, subgraph_copy)
```

Comment:
A detailed explanation of how and why this works should be added to the final version of the function.

Reply:
Done